### PR TITLE
Implement safety guards for TriggerProbes

### PR DESCRIPTION
**Pull request by evanchooly**
# What Does This Do

This PR introduces rate limiting, probe conditions, cool down support for trigger probes.

# Motivation

# Additional Notes

# Contributor Checklist

- [ ] Format the title [according the contribution guidelines](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#title-format)
- [ ] Assign the `type:` and (`comp:` or `inst:`) labels in addition to [any usefull labels](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#labels)
- [ ] Squash your commits prior merging or merge using GitHub's [Squash and merge](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squash-and-merge-your-commits)
- [ ] Don't use `close`, `fix` or any [linking keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) when referencing an issue.  
  Use `solves` instead, and assign the PR [milestone](https://github.com/DataDog/dd-trace-java/milestones) to the issue
- [ ] Update the [public documentation](https://docs.datadoghq.com/tracing/trace_collection/library_config/java/) in case of new configuration flag or behavior

Jira ticket: [DEBUG-2953 DEBUG-2954 DEBUG-2720]

[DEBUG-2953]: https://datadoghq.atlassian.net/browse/DEBUG-2953?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/TriggerProbe.java [hunk 1]
```diff
@@ -20,14 +24,26 @@
...
+  public TriggerProbe(
+      String language,
+      ProbeId probeId,
+      String[] tagStrs,
+      Where where,
+      ProbeCondition probeCondition,
+      Sampling sampling) {
+    super(language, probeId, tagStrs, where, MethodLocation.ENTRY);
+    this.probeCondition = probeCondition;
+    this.sampling = sampling;
```
Comment (evanchooly) [inline dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/TriggerProbe.java:46]: need to explore the sampling and tests a bit more to be sure it's doing what I think it is.

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/TriggerProbe.java [hunk 2]
```diff
@@ -2,32 +2,48 @@
...
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-public class TriggerProbe extends ProbeDefinition {
+public class TriggerProbe extends ProbeDefinition implements Sampled {
   private static final Logger LOGGER = LoggerFactory.getLogger(TriggerProbe.class);
 
+  private ProbeCondition probeCondition;
+
+  private Sampling sampling;
```
Comment (jpbempel) [inline dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/TriggerProbe.java:29]: I think we need to introduce a new class TriggerSampling based on
```
"triggerSampling": {
   "eventsPerSecond": 1,
   "coolDownInSeconds": 1
}
```

Comment (evanchooly) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/TriggerProbe.java:29]: 🤔   I'll see what the RC updates look like.

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/TriggerProbe.java [hunk 3]
```diff
@@ -1,33 +1,63 @@
...
+public class TriggerProbe extends ProbeDefinition implements Sampled {
   private static final Logger LOGGER = LoggerFactory.getLogger(TriggerProbe.class);
 
+  private static final int DEFAULT_COOLDOWN = 10;
+
+  private ProbeCondition probeCondition;
+
+  private Sampling sampling;
+
+  private long nextExecution = 0;
```
Comment (jpbempel) [inline dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/TriggerProbe.java:35]: would mark `nextExecution` `volatile`
multiple threads can execute the probe evaluation concurrently

File: dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/trigger/TriggerProbeTest.java [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+              .addTriggerProbes(singletonList(probe1))
+              .build();
+      installProbes(config);
+      Class<?> testClass = compileAndLoadClass(className);
+      int runs = 10000;
+      for (int i = 0; i < runs; i++) {
+        Reflect.onClass(testClass).call("main", "").get();
+      }
+
+      assertTrue(sampler.getCallCount() == 1);
```
Comment (jpbempel) [inline dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/trigger/TriggerProbeTest.java:76]: nit: assertEquals(1, sampler.getCallCount());

**Review by jpbempel — APPROVED** [review]

**Review by jpbempel — APPROVED** [review]

**Pull request merged**